components/mongodb/mongodb.app.mjs

```diff
@@ -1,11 +1,234 @@
+import mongoose from "mongoose";
+
 export default {
   type: "app",
   app: "mongodb",
-  propDefinitions: {},
+  propDefinitions: {
+    collection: {
+      label: "Collection",
+      type: "string",
+      description: "Collection where the new document will be inserted.",
+      async options() {
+        return this.getCollections();
+      },
+    },
+    document: {
+      label: "Document",
+      type: "string",
+      description: "The document ID",
+    },
+  },
   methods: {
-    // this.$auth contains connected account data
-    authKeys() {
-      console.log(Object.keys(this.$auth));
+    async connect() {
+      const {
+        username,
+        password,
+        database,
+        hostname,
+      } = this.$auth;
+      const uri = `mongodb+srv://${username}:${password}@${hostname}/${database}?retryWrites=true&w=majority`;
+      await mongoose.connect(uri);
+    },
+    getCollections() {
+      return new Promise((resolve, reject) => {
+        this.connect()
+          .then(async () => {
+            try {
+              const collections = await mongoose.connection.db.listCollections().toArray();
+              resolve(collections.map((collection) => collection.name).sort());
+            } catch (err) {
+              reject(err);
+            } finally {
+              mongoose.connection.close();
+            }
+          })
+          .catch((err) => reject(err));
+      });
+    },
+    createDocument(data, collection, parseNumbers, parseBooleans, parseDates) {
+      return new Promise((resolve, reject) => {
+        this.connect().then(async () => {
+          try {
+            const Model = mongoose.model(collection, this.getSchemaByData(data));
+            const document = new Model(this.parseStrings(
+              data,
+              parseNumbers,
+              parseBooleans,
+              parseDates,
+            ));
+            await document.save();
+            mongoose.connection.close(() => {
+              setTimeout(() => resolve(document));
+            });
+          } catch (err) {
+            mongoose.connection.close(() => {
+              setTimeout(() => reject(err));
+            });
+          }
+        })
+          .catch((err) => reject(err));
+      });
+    },
+    updateDocument(collection, _id, data, parseNumbers, parseBooleans, parseDates) {
+      return new Promise((resolve, reject) => {
+        this.connect().then(async () => {
+          try {
+            const Model = mongoose.model(collection, this.getSchemaByData(data));
+            await Model.findByIdAndUpdate(_id, this.parseStrings(
+              data,
+              parseNumbers,
+              parseBooleans,
+              parseDates,
+            ));
+            mongoose.connection.close(() => {
```

Comment:
Re: https://github.com/PipedreamHQ/pipedream/pull/2072#issuecomment-1015665877
> ACTIVE_HANDLEThis step was still trying to run code when the step ended. Make sure you promisify callback functions and await all Promises. (Reason: TLSSocket, Learn more: https://pipedream.com/docs/workflows/steps/code/async)

Changing `mongoose.connection.close` to [`mongoose.disconnect`](https://mongoosejs.com/docs/4.x/docs/api.html#index_Mongoose-disconnect) seemed to fix the issue for me.

Reply:
Unfortunately, it stills if I use disconnect. Sometimes I get the message, sometimes not. The only way I found to not get this error using `close` or `disconnect` was using a timeout :/